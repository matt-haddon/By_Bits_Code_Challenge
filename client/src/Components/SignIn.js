import React, { useState } from 'react';
import { useHistory } from 'react-router-dom';
import { TextField, Button } from '@material-ui/core';
import { postUser, userPolicy } from '../apiService';
import Empty from '../Assets/sun.jpg';

const SignIn = ({ setData, error, setError }) => {
    const [username, setUsername] = useState('');
    const [password, setPassword] = useState('');
    const history = useHistory();

    const handleSubmit = async (e) => {
        e.preventDefault();

        let token;

        if (username && password) {
            let res = await postUser(username, password);
            token = res.access_token;
            let result = await userPolicy(token);
            setData(result);
            history.push('/MyPolicy');
        } else {
            setError(true);
        }

        e.target.reset();
        setUsername('');
        setPassword('');
    };

    return (
        <div className="sign-in">
            <section className="picture">
                <img src={Empty} alt="login" />
            </section>
            <section className="formSection">
                <div className="formContainer">
                    <h1>Sign In</h1>
                    <div className="line"></div>
                    <form onSubmit={handleSubmit} id="form">
                        <TextField
                            label="Username"
                            id="username-input"
                            name="username"
                            type="text"
                            onChange={(e) => setUsername(e.target.value)}
                        />

                        <TextField
                            id="password-input"
                            label="Password"
                            type="password"
                            name="password"
                            onChange={(e) => setPassword(e.target.value)}
                        />
                        <Button
                            data-testid="submit"
                            color="primary"
                            type="submit"
                            id="submit"
                        >
                            Sign In
                        </Button>
                    </form>
                    <p className="warning">
                        {error
                            ? '❗️ Username or password not recognised. Please try again.'
                            : ''}
                    </p>
                </div>
            </section>
        </div>
    );
};

export default SignIn;
