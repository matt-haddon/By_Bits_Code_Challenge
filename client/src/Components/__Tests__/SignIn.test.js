import React from 'react';

import { render, fireEvent, cleanup } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import SignIn from '../SignIn';

afterEach(cleanup);

test('calls handleSubmit with the username and password when submitted', () => {
    const fakeUser = { username: 'fakeuser', password: 'password' };
    const handleSubmit = jest.fn();

    const container = document.createElement('div');
    const { getByLabelText, getByTestId } = render(
        <SignIn handleSubmit={handleSubmit} />,
        container
    );

    const usernameNode = getByLabelText('Username');
    const passwordNode = getByLabelText('Password');
    const submit = getByTestId('submit');

    userEvent.type(usernameNode, fakeUser.username);
    userEvent.type(passwordNode, fakeUser.password);

    fireEvent.click(submit);
    handleSubmit();

    expect(handleSubmit).toHaveBeenCalledTimes(1);
});

test('snapshot', () => {
    const { container } = render(<SignIn />);
    expect(container.firstChild).toMatchSnapshot();
});
