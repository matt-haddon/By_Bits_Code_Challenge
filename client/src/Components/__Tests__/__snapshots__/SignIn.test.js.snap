// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div
  class="sign-in"
>
  <section
    class="picture"
  >
    <img
      alt="login"
      src="sun.jpg"
    />
  </section>
  <section
    class="formSection"
  >
    <div
      class="formContainer"
    >
      <h1>
        Sign In
      </h1>
      <div
        class="line"
      />
      <form
        id="form"
      >
        <div
          class="MuiFormControl-root MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
            data-shrink="false"
            for="username-input"
            id="username-input-label"
          >
            Username
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiInput-input"
              id="username-input"
              name="username"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
            data-shrink="false"
            for="password-input"
            id="password-input-label"
          >
            Password
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiInput-input"
              id="password-input"
              name="password"
              type="password"
              value=""
            />
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
          data-testid="submit"
          id="submit"
          tabindex="0"
          type="submit"
        >
          <span
            class="MuiButton-label"
          >
            Sign In
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </form>
      <p
        class="warning"
      />
    </div>
  </section>
</div>
`;
